### ISSUE TITLE
Seeking Guidance on using the embeddings of megaDNA

### ISSUE THREAD

**Issue by YusenHou519** [lingxusb/megaDNA#4]
Dear Prof. Shao,

Thank you very much for sharing your excellent work on megaDNA.
I am sorry to trouble you again.

I am currently attempting to reproduce the results shown in Figure 1f (the infA single-codon mutation prediction). However, my experimental results are lower than those reported in the paper. I believe this may be due to some mistakes or misunderstandings in my implementation, and I would be very grateful for your kind guidance on the correct procedure.

## My reproduction steps
1. Downloaded the infA sequence from NCBI following the method described in the paper: https://www.ncbi.nlm.nih.gov/nuccore/NC_000913.3?report=fasta&from=926225&to=926443&strand=true and Downloaded the infA single-codon mutation data from the Cell supplementary material: https://www.cell.com/cms/10.1016/j.cels.2016.11.004/attachment/6dccc23f-92df-4cab-9953-e6af34fa52a1/mmc5.zip

2. Used the pretrained model megaDNA_phage_145M.pt to obtained the three transformer-layer embeddings from the model output. Following the suggestion in Issue #2, applied average pooling across the intermediate dimensions of each layer’s output and then concatenated the three pooled embeddings to form the sequence representation.

3. Used sklearn.linear_model.RidgeCV for fitting, with 5-fold cross-validation and Spearman correlation as the evaluation metric.

However, the Spearman (0.6852) and Pearson (0.7894) correlations I obtained (on all **4672 datapoints**) are both significantly lower than the values reported in the paper.

I suspect that some implementation details in my reproduction may not fully match those used in your experiments, and I would sincerely appreciate your guidance on the correct implementation.

Below is my implementation code:

```
import torch
import pickle
import numpy as np
import pandas as pd
from tqdm import tqdm
import warnings 

RECALCULATE_EMBEDS=False

def megadna_encode_sequence(sequence, nt_vocab=['**', 'A', 'T', 'C', 'G', '#']):
    """Encode a DNA sequence to its numerical representation."""
    return [0] + [nt_vocab.index(nucleotide) if nucleotide in nt_vocab else 1 for nucleotide in sequence] + [5]
    # I try to remove the (start token & end token) or start token, but the results are worse.
    # return [nt_vocab.index(nucleotide) if nucleotide in nt_vocab else 5 for nucleotide in sequence] + [5]

def megadna_get_embed_for_sequence(model, sequence, device):
    """Get model embed for a given sequence."""
    input_seq = torch.tensor(sequence).unsqueeze(0).to(device)
    with torch.inference_mode():
        embed = model(input_seq, return_value='embedding')
    return embed

def pool_hidden_states(hidden_states, strategy="last_mean"):
    """Pool hidden states into a single feature vector per sequence."""
    # hidden_states: list of tensors [batch=1, seq_len, hidden_dim]
    if strategy == "last_mean":
        last = hidden_states[-1]
        embed = last.detach().mean(dim=(0,1))  # [1, hidden_dim]
        return embed
    elif strategy == "all_concat_mean":
        pooled_per_layer = [i.detach().mean(dim=(0,1)) for i in hidden_states]
        embed = torch.cat(pooled_per_layer, dim=-1) 
        return embed # [1, sum(d_l)]
    else:
        raise ValueError(f"Unknown pooling strategy: {strategy}")
    
if RECALCULATE_EMBEDS:
    model_path = "megaDNA_phage_145M.pt"
    device = 'cuda:0' if torch.cuda.is_available() else 'cpu' 
    model = torch.load(model_path, map_location=torch.device(device), weights_only=False).eval() 

    mut_seqs = pd.read_csv("mutants_sequences.tsv", sep='\t')
    X = []
    Y = []
    for idx, row in tqdm(mut_seqs.iterrows(), total=len(mut_seqs)):
        encoded_sequence = megadna_encode_sequence(row['seq'])
        output = megadna_get_embed_for_sequence(model, encoded_sequence, device)
        output = [i.cpu() for i in output]
        X.append(pool_hidden_states(output, strategy="all_concat_mean"))
        Y.append(torch.tensor(row['fitness']))
    with open("megaDNA_train_ridgecv_data.pkl", "wb") as f:
        pickle.dump(((torch.stack(X)).numpy(), (torch.stack(Y)).numpy()), f)
        
# load data
import pickle
with open("megaDNA_train_ridgecv_data.pkl", "rb") as f:
    X, Y = pickle.load(f)
print(f"X.shape: {X.shape}")
print(f"Y.shape: {Y.shape}")

from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import RidgeCV
from sklearn.model_selection import KFold, cross_validate
from sklearn.metrics import make_scorer
from scipy.stats import spearmanr
import numpy as np

def _spearman_corr(y_true, y_pred):
    rho, _ = spearmanr(y_pred, y_true)
    return 0.0 if np.isnan(rho) else float(rho)

def _pearson_corr(y_true, y_pred):
    r = np.corrcoef(y_pred, y_true)[0, 1]
    return 0.0 if np.isnan(r) else float(r)

outer_cv = KFold(n_splits=5, shuffle=True, random_state=42)
inner_cv = KFold(n_splits=5, shuffle=True, random_state=42)

spearman_scorer = make_scorer(_spearman_corr, greater_is_better=True)
pearson_scorer = make_scorer(_pearson_corr, greater_is_better=True)

model = make_pipeline(
    StandardScaler(with_mean=True, with_std=True),
    RidgeCV(alphas=np.logspace(-3, 3, 100), cv=inner_cv, scoring=spearman_scorer)
)

cv_results = cross_validate(
    model, X, Y, cv=outer_cv,
    scoring={"spearman": spearman_scorer, "pearson": pearson_scorer},
    n_jobs=-1, return_estimator=True
)
print(f"Spearman: {cv_results['test_spearman']}")
print(f"Mean Spearman: {np.mean(cv_results['test_spearman']):.4f} ± {np.std(cv_results['test_spearman']):.4f}")
print(f"Pearson:  {cv_results['test_pearson']}")
print(f"Mean Pearson:  {np.mean(cv_results['test_pearson']):.4f} ± {np.std(cv_results['test_pearson']):.4f}")

selected_alphas = [
    est.named_steps['ridgecv'].alpha_ for est in cv_results['estimator']
]
print("Selected alphas per fold:", selected_alphas)
```

Thank you very much for taking the time to read this message, and I am deeply grateful for any clarification or reference pseudocode you could kindly provide.

Best regards,
Yusen

**Comment by lingxusb** [comment]
Dear Yusen,

My apologies for the very late reply, and thank you very much for your interest in our work on megaDNA and for taking the time to reproduce the infA mutation results in such detail.

Regarding Figure 1f: the Spearman correlation in that panel is around 0.7, so your value of 0.6852 is consistent with what we obtained. Small differences are within the variation across different random seeds or cross-validation splits, so I believe your reproduction looks good.

Thank you again for your careful work and for reaching out. Please feel free to contact me anytime if further questions come up.

Best regards,
Bin

**Comment by YusenHou519** [comment]
Dear Prof. Shao,

Thank you very much for your detailed and kind reply!
Your explanation resolved my confusion and was extremely helpful to my understanding. I truly appreciate the time and effort you took to clarify this issue.

I had the pleasure of attending your talk at the genomics conference in Dapeng, Shenzhen — your presentation left a deep impression on me. Your patience and dedication to research deeply impress me, and it’s inspiring to see such scientific rigor and generosity in sharing your expertise.

I sincerely hope to have more opportunities to learn from you in the future.

Thank you again for your guidance!

Best regards,
Yusen
